pkg/mutation/mutation_function_test.go

```diff
@@ -0,0 +1,234 @@
+package mutation_test
+
+import (
+	"fmt"
+	"testing"
+
+	mutationsv1alpha1 "github.com/open-policy-agent/gatekeeper/apis/mutations/v1alpha1"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const TestValue = "testValue"
+
+func prepareTestPod(t *testing.T) *unstructured.Unstructured {
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "testpod1",
+			Namespace: "foo",
+			Labels:    map[string]string{"a": "b"},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:  "testname1",
+					Ports: []corev1.ContainerPort{{Name: "portName1"}},
+				},
+				{
+					Name: "testname2",
+					Ports: []corev1.ContainerPort{
+						{Name: "portName2A"},
+						{Name: "portName2B"},
+					},
+				},
+				{
+					Name:  "testname3",
+					Ports: []corev1.ContainerPort{{Name: "portName3"}},
+				},
+			},
+		},
+	}
+	podObject, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	if err != nil {
+		t.Errorf("Failed to convert pod to unstructured %v", err)
+	}
+	return &unstructured.Unstructured{Object: podObject}
+}
+
+func TestObjects(t *testing.T) {
+
+	testFunc := func(unstructured *unstructured.Unstructured) {
+		labels := unstructured.Object["metadata"].(map[string]interface{})["labels"]
+		if labels.(map[string]interface{})["labelA"] != TestValue {
+			t.Errorf("Failed to update pod")
+		}
+
+	}
+
+	testMutation(
+		"metadata.labels.labelA",
+		TestValue,
+		prepareTestPod(t),
+		testFunc,
+		t,
+	)
+}
+
+func TestObjectsAndLists(t *testing.T) {
+	testFunc := func(unstructured *unstructured.Unstructured) {
+		containers := unstructured.Object["spec"].(map[string]interface{})["containers"]
+		for _, container := range containers.([]interface{}) {
+			containerAsMap := container.(map[string]interface{})
+			if containerAsMap["name"] == "testname2" {
+				ports := containerAsMap["ports"]
+				for _, port := range ports.([]interface{}) {
+					portAsMap := port.(map[string]interface{})
+					if portAsMap["name"] == "portName2B" {
+						if portAsMap["hostIP"] != TestValue {
+							t.Errorf("Failed to update pod")
+						}
+					} else {
+						if _, ok := port.(map[string]interface{})["hostIP"]; ok {
+							t.Errorf("Unexpected pod was updated")
+						}
+					}
+				}
+			} else {
+				for _, port := range container.(map[string]interface{})["ports"].([]interface{}) {
+					if _, ok := port.(map[string]interface{})["hostIP"]; ok {
+						t.Errorf("Unexpected pod was updated")
+					}
+				}
+			}
+		}
+	}
+
+	testMutation(
+		`spec.containers["name": "testname2"].ports["name": "portName2B"].hostIP`,
```

Comment:
I'm not 100% sure this test belongs here, but are we enforcing somewhere that we can't have a location like:

``spec.containers["name": "foo"].ports["name": "portName"].name` ?

Where we are changing the value of the object's key field?

Reply:
We don't have this here, but I would say this should be added to parser tests.
I'll add new PR for this.

Reply:
+1 to having it elsewhere @shomron may have an opinion as to whether it belongs in the parser test.

Not blocking this PR IMO, I filed an issue making sure we double check this invariant is enforced.